geosegmentize: prevent OOMs

otan opened the pull request:
geosegmentize can be given inputs that break a line down into too many
coordinates. Guard against this by ensuring a maximum amount of
coordinates allowed between points.

PostgreSQL errors the allocate here too.

Resolves #52218.
Resolves #52217.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52261)
<!-- Reviewable:end -->

sumeerbhola reviewed (commented):
> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52261)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geogfn/segmentize.go, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52261#-MDtxDz1ATTKLOfwvNMd:-MDtxDz1ATTKLOfwvNMe:bgtuf35) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c95911bb1376b8f1a966324ba74e4b870584ca5/pkg/geo/geogfn/segmentize.go#L67)):*
> ```Go
> 	// This calculation is to determine the total number of segment between given
> 	// 2 coordinates, ensuring that the segments are divided into parts divisible by
> 	// a power of 2.
> ```

why does this need to be a power of 2.
___
*[pkg/geo/geosegmentize/geosegmentize.go, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52261#-MDtxjJH53NdIwwBp0zP:-MDtxjJH53NdIwwBp0zQ:b-7tys0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c95911bb1376b8f1a966324ba74e4b870584ca5/pkg/geo/geosegmentize/geosegmentize.go#L19)):*
> ```Go
> 
> // MaxPoints is the maximum number of points segmentize is allowed to generate.
> const MaxPoints = 16336
> ```

is this the same limit value as PostGIS?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52261)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geogfn/segmentize.go, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52261#-MDtxDz1ATTKLOfwvNMd:-MDtyMGn4UCvW7M63gMY:bq8tqte) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c95911bb1376b8f1a966324ba74e4b870584ca5/pkg/geo/geogfn/segmentize.go#L67)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why does this need to be a power of 2.
</blockquote></details>

It's how the algorithm in PostGIS works.
___
*[pkg/geo/geosegmentize/geosegmentize.go, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52261#-MDtxjJH53NdIwwBp0zP:-MDtyPDlAi6aN9a8Ms0n:b-gy2i9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c95911bb1376b8f1a966324ba74e4b870584ca5/pkg/geo/geosegmentize/geosegmentize.go#L19)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is this the same limit value as PostGIS?
</blockquote></details>

PostGIS seemingly uses some allocator in PostgreSQL, which instead reports an OOM if it gets too big. We don't do something similar, so hardcoded something insanely big for now.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52261#-:-MDuDZZ1AxuR_W6114kJ:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geogfn/segmentize.go, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52261#-MDtxDz1ATTKLOfwvNMd:-MDuD4Vh0iOpjZZXDsUI:b27o9kh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c95911bb1376b8f1a966324ba74e4b870584ca5/pkg/geo/geogfn/segmentize.go#L67)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

It's how the algorithm in PostGIS works.
</blockquote></details>

Can you add a comment stating that since otherwise the reader is left wondering why not `n := ceil(chordAngleBetweenPoints/segmentMaxAngle)`


<!-- Sent from Reviewable.io -->

otan committed:
geosegmentize: prevent OOMs

geosegmentize can be given inputs that break a line down into too many
coordinates. Guard against this by ensuring a maximum amount of
coordinates allowed between points.

PostgreSQL errors the allocate here too.

Release note: None

otan commented:
thanks!

bors r=sumeerbhola
